feat: User friendly error when executable not found

jblagden opened the pull request:
Fixes: https://github.com/OpenJobDescription/openjd-adaptor-runtime-for-python/issues/113

### What was the problem/requirement? (What/Why)

When the adaptor fails to find the set executable, it'll fail with `[WinError 2] The system cannot find the file specified`, which without any context doesn't set users up to sort the issue out.

### What was the solution? (How)

This change adds user friendly messages about what's failing and possible reasons why.

### What is the impact of this change?

Users will see more logging when the adaptor fails to find the executable it has been set to run.

### How was this change tested?

Currently a gap in this PR - I'm not sure where to start testing if a file does or does not exist and modifying the environment in a test.

See [DEVELOPMENT.md](https://github.com/OpenJobDescription/openjd-adaptor-runtime-for-python/blob/mainline/DEVELOPMENT.md#testing) for information on running tests.

- Have you run the unit tests?

Yes, to failures on TestIntegrationClientInterface, TestIntegrationLoggingSubprocess, and TestDaemonMode:
```
FAILED test/openjd/adaptor_runtime_client/integ/test_integration_client_interface.py::TestIntegrationClientInterface::test_graceful_shutdown - AssertionError: assert 'Received SIGTERM signal.' in ''
FAILED test/openjd/adaptor_runtime/integ/process/test_integration_logging_subprocess.py::TestIntegrationLoggingSubprocess::test_stop_process[StopProcessWhenSIGTERMFails] - Failed: Timeout >5.0s
FAILED test/openjd/adaptor_runtime/integ/background/test_background_mode.py::TestDaemonMode::test_shutdown - psutil.TimeoutExpired: timeout after 1 seconds (pid=11312)
```

I'm working in the Windows subsystem for Linux which might be the cause of the failing timeouts but I'm not sure about the SIGTERM issue. I'm hoping the github runners shed a little light on what's happening.

This still needs a test for the case when an alias is set but the executable isn't found. I'm not certain that it's a usefully different code-path. Removing that may be better.

### Was this change documented?

Yes.

### Is this a breaking change?

No.

### Does this change impact security?

No.

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

dependabot[bot] committed:
chore(deps): update python-semantic-release requirement (#159)

Updates the requirements on [python-semantic-release](https://github.com/python-semantic-release/python-semantic-release) to permit the latest version.
- [Release notes](https://github.com/python-semantic-release/python-semantic-release/releases)
- [Changelog](https://github.com/python-semantic-release/python-semantic-release/blob/master/CHANGELOG.md)
- [Commits](https://github.com/python-semantic-release/python-semantic-release/compare/v9.12...v9.14)

---
updated-dependencies:
- dependency-name: python-semantic-release
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

joel-wong-aws committed:
chore: allow pywin32 >= 307 (#162)

Signed-off-by: Joel Wong <127782171+joel-wong-aws@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

client-software-ci committed:
chore(release): 0.8.2 (#165)

Signed-off-by: client-software-ci <129794699+client-software-ci@users.noreply.github.com>
Signed-off-by: Joel Wong <127782171+joel-wong-aws@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

karthikbekalp committed:
fix!: Ensure all open uses UTF-8 instead of default encoding. (#171)

Signed-off-by: Karthik Bekal Pattathana <133984042+karthikbekalp@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

client-software-ci committed:
chore(release): 0.9.0 (#173)

Signed-off-by: client-software-ci <129794699+client-software-ci@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

jblagden committed:
feat! Adds user-friendly messaging when the adaptor executable isn't found

Also checks the result of `which`, just in case there's an alias that's pointing to nowhere.

Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

jblagden committed:
Adds user-friendly messaging when the adaptor executable isn't found.

Also checks the result of `which`, just in case there's an alias that's pointing to nowhere.

jblagden committed:
feat! Adds user-friendly messaging when the adaptor executable isn't found

Also checks the result of `which`, just in case there's an alias that's pointing to nowhere.

Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

jblagden committed:
Merge branch 'better_error_subprocess_executable' of https://github.com/jblagden/openjd-adaptor-runtime-for-python into better_error_subprocess_executable

joel-wong-aws committed:
chore: update GitHub issue templates (#176)

Signed-off-by: Joel Wong <127782171+joel-wong-aws@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

jblagden committed:
Resolves code smells called out by sonarqube

Fixed an always-true type check

Removed an explicit throw of un-caught exception types

Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

jblagden committed:
Resolves code smells called out by sonarqube

Fixed an always-true type check

Removed an explicit throw of un-caught exception types

joel-wong-aws committed:
chore: increase log level for daemon _serve args to info (#169)

Signed-off-by: Joel Wong <127782171+joel-wong-aws@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

jblagden committed:
Adds a test for executable not found error

Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

jblagden committed:
Adds a test for executable not found error

jblagden committed:
Adds a test for executable not found error

jblagden committed:
Adds a test for executable not found error

Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

jblagden committed:
Merge branch 'better_error_subprocess_executable' of https://github.com/jblagden/openjd-adaptor-runtime-for-python into better_error_subprocess_executable

jblagden committed:
Merge branch 'better_error_subprocess_executable' of https://github.com/jblagden/openjd-adaptor-runtime-for-python into better_error_subprocess_executable

jblagden committed:
Merge branch 'mainline' into better_error_subprocess_executable

mwiebe commented on src/openjd/adaptor_runtime/process/_logging_subprocess.py:
```diff
@@ -64,7 +69,26 @@ def __init__(
+        except FileNotFoundError as fnf_error:
+            # In ManagedProcess we prepend the executable to the list of arguments before creating a LoggingSubprocess
+            executable = args[0]
+            exe_path = shutil.which(executable)
```
If we're using `shutil.which` to resolve the program, it will be better to do this before we call `subprocess.Popen`, and then modify `args[0]` to contain the result of that. We found that on Windows, `subprocess.Popen` and `shutil.which` can get different answers, so the approach you've implemented could make inaccurate error messages.

jblagden replied:
Interesting - do you recall how we might be able to re-create that? It might be worth a unit test to save future folks from a similar pitfall. I'll get a tweak up as soon as I can - likely tomorrow morning.

jblagden replied:
What a mistake it was to commit to a time I'd get back to this!

I think I've got it all straightened out now though, let me know what you think when you have a chance.

joel-wong-aws committed:
chore: add maintenance label to maintenance GitHub template (#182)

Signed-off-by: Joel Wong <127782171+joel-wong-aws@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update python-semantic-release requirement (#181)

Updates the requirements on [python-semantic-release](https://github.com/python-semantic-release/python-semantic-release) to permit the latest version.
- [Release notes](https://github.com/python-semantic-release/python-semantic-release/releases)
- [Changelog](https://github.com/python-semantic-release/python-semantic-release/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/python-semantic-release/python-semantic-release/compare/v9.14...v9.20)

---
updated-dependencies:
- dependency-name: python-semantic-release
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Charles Moore <122481442+moorec-aws@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update hatch requirement from ==1.13.* to ==1.14.* (#168)

Updates the requirements on [hatch](https://github.com/pypa/hatch) to permit the latest version.
- [Release notes](https://github.com/pypa/hatch/releases)
- [Commits](https://github.com/pypa/hatch/compare/hatch-v1.13.0...hatch-v1.14.0)

---
updated-dependencies:
- dependency-name: hatch
  dependency-type: direct:development
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Charles Moore <122481442+moorec-aws@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update ruff requirement from ==0.8.* to ==0.9.* (#175)

Updates the requirements on [ruff](https://github.com/astral-sh/ruff) to permit the latest version.
- [Release notes](https://github.com/astral-sh/ruff/releases)
- [Changelog](https://github.com/astral-sh/ruff/blob/main/CHANGELOG.md)
- [Commits](https://github.com/astral-sh/ruff/compare/0.8.0...0.9.1)

---
updated-dependencies:
- dependency-name: ruff
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Charles Moore <122481442+moorec-aws@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

jblagden committed:
Merge branch 'better_error_subprocess_executable' of https://github.com/jblagden/openjd-adaptor-runtime-for-python into better_error_subprocess_executable

jblagden committed:
Updates test assertions to include the whole executable path

Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

moorec-aws committed:
test: Broaden regex to ignore copyright header file created by setuptools (#186)

Signed-off-by: Charles Moore <122481442+moorec-aws@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update psutil requirement from ==6.1.* to ==7.0.* (#185)

Updates the requirements on [psutil](https://github.com/giampaolo/psutil) to permit the latest version.
- [Changelog](https://github.com/giampaolo/psutil/blob/master/HISTORY.rst)
- [Commits](https://github.com/giampaolo/psutil/compare/release-6.1.0...release-7.0.0)

---
updated-dependencies:
- dependency-name: psutil
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Charles Moore <122481442+moorec-aws@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update black requirement from ==24.* to ==25.* (#184)

Updates the requirements on [black](https://github.com/psf/black) to permit the latest version.
- [Release notes](https://github.com/psf/black/releases)
- [Changelog](https://github.com/psf/black/blob/main/CHANGES.md)
- [Commits](https://github.com/psf/black/compare/24.1a1...25.1.0)

---
updated-dependencies:
- dependency-name: black
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update python-semantic-release requirement (#183)

Updates the requirements on [python-semantic-release](https://github.com/python-semantic-release/python-semantic-release) to permit the latest version.
- [Release notes](https://github.com/python-semantic-release/python-semantic-release/releases)
- [Changelog](https://github.com/python-semantic-release/python-semantic-release/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/python-semantic-release/python-semantic-release/compare/v9.20...v9.21)

---
updated-dependencies:
- dependency-name: python-semantic-release
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update ruff requirement from ==0.9.* to ==0.11.* (#189)

Updates the requirements on [ruff](https://github.com/astral-sh/ruff) to permit the latest version.
- [Release notes](https://github.com/astral-sh/ruff/releases)
- [Changelog](https://github.com/astral-sh/ruff/blob/main/CHANGELOG.md)
- [Commits](https://github.com/astral-sh/ruff/compare/0.9.0...0.11.0)

---
updated-dependencies:
- dependency-name: ruff
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update pytest-cov requirement from ==6.0.* to ==6.1.* (#190)

Updates the requirements on [pytest-cov](https://github.com/pytest-dev/pytest-cov) to permit the latest version.
- [Changelog](https://github.com/pytest-dev/pytest-cov/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/pytest-dev/pytest-cov/compare/v6.0.0...v6.1.1)

---
updated-dependencies:
- dependency-name: pytest-cov
  dependency-version: 6.1.1
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

edwards-aws committed:
chore: changes for mypy update (#191)

Signed-off-by: Cody Edwards <edwards@amazon.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update pytest-timeout requirement from ==2.3.* to ==2.4.* (#192)

Updates the requirements on [pytest-timeout](https://github.com/pytest-dev/pytest-timeout) to permit the latest version.
- [Commits](https://github.com/pytest-dev/pytest-timeout/compare/2.3.0...2.4.0)

---
updated-dependencies:
- dependency-name: pytest-timeout
  dependency-version: 2.4.0
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

moorec-aws committed:
chore(deps): update pytest to 8.3.* (#193)

Signed-off-by: Charles Moore <122481442+moorec-aws@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update hatch-vcs requirement from ==0.4.* to ==0.5.*

Updates the requirements on [hatch-vcs](https://github.com/ofek/hatch-vcs) to permit the latest version.
- [Release notes](https://github.com/ofek/hatch-vcs/releases)
- [Changelog](https://github.com/ofek/hatch-vcs/blob/master/HISTORY.md)
- [Commits](https://github.com/ofek/hatch-vcs/compare/v0.4.0...v0.5.0)

---
updated-dependencies:
- dependency-name: hatch-vcs
  dependency-version: 0.5.0
  dependency-type: direct:development
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update pytest-xdist requirement from ==3.6.* to ==3.7.*

Updates the requirements on [pytest-xdist](https://github.com/pytest-dev/pytest-xdist) to permit the latest version.
- [Release notes](https://github.com/pytest-dev/pytest-xdist/releases)
- [Changelog](https://github.com/pytest-dev/pytest-xdist/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/pytest-dev/pytest-xdist/compare/v3.6.0...v3.7.0)

---
updated-dependencies:
- dependency-name: pytest-xdist
  dependency-version: 3.7.0
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update pytest requirement from ==8.3.* to ==8.4.*

Updates the requirements on [pytest](https://github.com/pytest-dev/pytest) to permit the latest version.
- [Release notes](https://github.com/pytest-dev/pytest/releases)
- [Changelog](https://github.com/pytest-dev/pytest/blob/main/CHANGELOG.rst)
- [Commits](https://github.com/pytest-dev/pytest/compare/8.3.0.dev0...8.4.0)

---
updated-dependencies:
- dependency-name: pytest
  dependency-version: 8.4.0
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update mypy requirement from ==1.15.* to ==1.16.*

Updates the requirements on [mypy](https://github.com/python/mypy) to permit the latest version.
- [Changelog](https://github.com/python/mypy/blob/master/CHANGELOG.md)
- [Commits](https://github.com/python/mypy/compare/v1.15.0...v1.16.0)

---
updated-dependencies:
- dependency-name: mypy
  dependency-version: 1.16.0
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

epmog committed:
fix: sdist failed to install (#199)

Signed-off-by: Morgan Epp <60796713+epmog@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

mwiebe committed:
docs: Refine the introduction, move usage into docs dir

* Changes to the introduction are to clarify the purpose of this
  library, making it easier to understand what features it provides.
* Moved the content about using the library from README.md into a docs
  directory. With the re-organization, there's a place to add new
  content like worked examples or more details about interface
  features the library supports.

Signed-off-by: Mark <399551+mwiebe@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

moorec-aws committed:
ci: update to tag based release workflows (#201)

Signed-off-by: Charles Moore <122481442+moorec-aws@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

mwiebe committed:
docs: Write a worked example of defining a Blender adaptor (#202)

This is a proposal to try a very different adaptor interface than
is currently used.

Signed-off-by: Mark <399551+mwiebe@users.noreply.github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): update python-semantic-release requirement

Updates the requirements on [python-semantic-release](https://github.com/python-semantic-release/python-semantic-release) to permit the latest version.
- [Release notes](https://github.com/python-semantic-release/python-semantic-release/releases)
- [Changelog](https://github.com/python-semantic-release/python-semantic-release/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/python-semantic-release/python-semantic-release/compare/v9.21...v10.0.2)

---
updated-dependencies:
- dependency-name: python-semantic-release
  dependency-version: 10.0.2
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

jblagden committed:
Merge branch 'OpenJobDescription:mainline' into better_error_subprocess_executable

jblagden committed:
Merge branch 'better_error_subprocess_executable' of https://github.com/jblagden/openjd-adaptor-runtime-for-python into better_error_subprocess_executable

Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

jblagden committed:
Merge branch 'better_error_subprocess_executable' of https://github.com/jblagden/openjd-adaptor-runtime-for-python into better_error_subprocess_executable

Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

jblagden committed:
Signing unsigned commits

Signed-off-by: Justin Blagden <jblagden@users.noreply.github.com>

jblagden committed:
Merge branch 'OpenJobDescription:mainline' into better_error_subprocess_executable
